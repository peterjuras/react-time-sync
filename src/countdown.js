import React, { Component } from "react";
import PropTypes from "prop-types";
import TimeContext, { TIMESYNC_PROP } from "./context";
import { hasConfigChanged } from "./config";

function getCountdownConfig(props) {
  const newCountdownConfig = {};

  if (props.interval) {
    newCountdownConfig.interval = props.interval;
  }

  if (props.until) {
    newCountdownConfig.until = props.until;
  } else {
    newCountdownConfig.until = 0;
  }

  return newCountdownConfig;
}

class Countdown extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    until: PropTypes.number,
    interval: PropTypes.string
  };

  static defaultProps = {
    until: null,
    interval: null
  };

  constructor(props) {
    super(props);

    if (!props[TIMESYNC_PROP]) {
      throw new Error(
        "Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"
      );
    }
  }

  state = {};

  componentDidMount() {
    this.resetCountdown();
  }

  componentDidUpdate(prevProps, prevState) {
    const { countdownConfig } = this.state;

    if (prevState.countdownConfig !== countdownConfig) {
      this.resetCountdown();
    }
  }

  componentWillUnmount() {
    if (this.stopCountdown) {
      this.stopCountdown();
    }
  }

  onCountdownTick = timeLeft => {
    this.setState({ timeLeft });
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    const countdownConfig = getCountdownConfig(nextProps);
    const hasChanged = hasConfigChanged(
      prevState.countdownConfig,
      countdownConfig
    );

    if (hasChanged) {
      return {
        countdownConfig,
        timeLeft: nextProps[TIMESYNC_PROP].getTimeLeft(countdownConfig)
      };
    }

    return null;
  }

  resetCountdown() {
    const { countdownConfig, timeLeft } = this.state;

    if (this.stopCountdown) {
      this.stopCountdown();
      this.stopCountdown = null;
    }

    if (timeLeft === 0) {
      return;
    }

    const { [TIMESYNC_PROP]: timeSync } = this.props;

    this.stopCountdown = timeSync.createCountdown(
      this.onCountdownTick,
      countdownConfig
    );
  }

  render() {
    const { children } = this.props;
    const { timeLeft } = this.state;

    return children({
      timeLeft
    });
  }
}

export default function CountdownWrapper(props) {
  return (
    <TimeContext.Consumer>
      {timeSync => <Countdown {...{ ...props, [TIMESYNC_PROP]: timeSync }} />}
    </TimeContext.Consumer>
  );
}
