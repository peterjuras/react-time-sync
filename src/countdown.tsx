import React, { Component, FunctionComponent, ReactElement } from "react";
import PropTypes from "prop-types";
import TimeContext, { TIMESYNC_PROP, TimeSyncContext } from "./context";
import { hasConfigChanged } from "./config";
import { Interval } from "time-sync/constants";

export interface CountdownConfig {
  until?: number;
  interval?: Interval;
}

export interface SafeCountdownConfig extends CountdownConfig {
  until: number;
  [key: string]: number | string | undefined;
}

export type CountdownChildrenType = (obj: {
  timeLeft: number;
}) => ReactElement | ReactElement[];

interface CountdownConfigProps extends CountdownConfig {
  children?: CountdownChildrenType;
}

interface CountdownProps extends CountdownConfigProps {
  [TIMESYNC_PROP]: TimeSyncContext;
}

function getCountdownConfig(props: CountdownProps): SafeCountdownConfig {
  const newCountdownConfig: SafeCountdownConfig = { until: 0 };

  if (props.interval) {
    newCountdownConfig.interval = props.interval;
  }

  if (props.until) {
    newCountdownConfig.until = props.until;
  }

  return newCountdownConfig;
}

interface CountdownState {
  countdownConfig: SafeCountdownConfig;
  timeLeft: number;
}

class Countdown extends Component<CountdownProps, CountdownState> {
  public static propTypes = {
    children: PropTypes.func.isRequired,
    until: PropTypes.number,
    interval: PropTypes.string
  };

  public static defaultProps = {
    until: null,
    interval: null
  };

  public static getDerivedStateFromProps(
    nextProps: CountdownProps,
    prevState: CountdownState
  ): CountdownState | null {
    const countdownConfig = getCountdownConfig(nextProps);
    const hasChanged = hasConfigChanged(
      prevState.countdownConfig,
      countdownConfig
    );

    if (hasChanged) {
      return {
        countdownConfig,
        timeLeft: nextProps[TIMESYNC_PROP].getTimeLeft(countdownConfig)
      };
    }

    return null;
  }

  public state: CountdownState = {
    countdownConfig: { until: 0 },
    timeLeft: 0
  };

  private stopCountdown?: (() => void) | null;

  public constructor(props: CountdownProps) {
    super(props);

    if (!props[TIMESYNC_PROP]) {
      throw new Error(
        "Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"
      );
    }
  }

  public componentDidMount(): void {
    this.resetCountdown();
  }

  public componentDidUpdate(
    _: CountdownProps,
    prevState: CountdownState
  ): void {
    const { countdownConfig } = this.state;

    if (prevState.countdownConfig !== countdownConfig) {
      this.resetCountdown();
    }
  }

  public componentWillUnmount(): void {
    if (this.stopCountdown) {
      this.stopCountdown();
    }
  }

  public render(): ReturnType<CountdownChildrenType> | null {
    const { children } = this.props;
    const { timeLeft } = this.state;

    if (!children) {
      return null;
    }

    return children({
      timeLeft
    });
  }

  private onCountdownTick = (timeLeft: number) => {
    this.setState({ timeLeft });
  };

  private resetCountdown(): void {
    const { countdownConfig, timeLeft } = this.state;

    if (this.stopCountdown) {
      this.stopCountdown();
      this.stopCountdown = null;
    }

    if (timeLeft === 0) {
      return;
    }

    const { [TIMESYNC_PROP]: timeSync } = this.props;

    this.stopCountdown = timeSync.createCountdown(
      this.onCountdownTick,
      countdownConfig
    );
  }
}

const CountdownWrapper: FunctionComponent<CountdownConfigProps> = (
  props
): JSX.Element => {
  return (
    <TimeContext.Consumer>
      {timeSync => <Countdown {...{ ...props, [TIMESYNC_PROP]: timeSync }} />}
    </TimeContext.Consumer>
  );
};
export default CountdownWrapper;
