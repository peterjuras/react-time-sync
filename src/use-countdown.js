import { useContext, useEffect, useState } from "react";
import TimeContext from "./context";

export function useCountdown(countdownConfig = {}) {
  const timeSync = useContext(TimeContext);
  if (!timeSync) {
    throw new Error(
      "Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"
    );
  }

  const usedCountdownConfig = {
    ...countdownConfig,
    until: countdownConfig.until || 0
  };

  const [timeLeft, setTimeLeft] = useState(
    timeSync.getTimeLeft(usedCountdownConfig)
  );

  useEffect(
    () => {
      const officialTimeLeft = timeSync.getTimeLeft(usedCountdownConfig);
      if (officialTimeLeft !== timeLeft) {
        setTimeLeft(officialTimeLeft);
      }

      let stopCountdown;
      if (officialTimeLeft > 0) {
        stopCountdown = timeSync.createCountdown(
          setTimeLeft,
          usedCountdownConfig
        );
      }
      return stopCountdown;
    },
    [usedCountdownConfig.until, usedCountdownConfig.interval]
  );

  return timeLeft;
}
