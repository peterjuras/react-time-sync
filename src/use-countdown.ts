import { useContext, useEffect, useDebugValue } from "react";
import TimeContext from "./context";
import { Interval } from "time-sync/constants";
import { useStateWithDeps } from "use-state-with-deps";

export interface PartialCountdownConfig {
  until?: number;
  interval?: Interval;
}

export interface CountdownConfig extends PartialCountdownConfig {
  until: number;
}

function getUsableConfig(
  countdownConfig: PartialCountdownConfig,
): CountdownConfig {
  return {
    ...countdownConfig,
    until: countdownConfig.until || 0,
  };
}

export function useCountdown(
  countdownConfig: PartialCountdownConfig = {},
): number {
  const timeSync = useContext(TimeContext);
  const usableConfig = getUsableConfig(countdownConfig);

  const [timeLeft, setTimeLeft] = useStateWithDeps(() => {
    return timeSync.getTimeLeft({
      interval: usableConfig.interval,
      until: usableConfig.until,
    });
  }, [usableConfig.interval, usableConfig.until]);

  useEffect((): (() => void) | void => {
    if (Date.now() < usableConfig.until) {
      return timeSync.createCountdown((newTimeLeft): void => {
        setTimeLeft(newTimeLeft);
      }, usableConfig);
    }
  }, [usableConfig.until, usableConfig.interval]);

  useDebugValue(timeLeft);
  return timeLeft;
}
