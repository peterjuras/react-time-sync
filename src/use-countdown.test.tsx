import React from "react";
import TimeSync from "time-sync";
import TimeContext from "./context";
import { useCountdown } from "./use-countdown";
import lolex from "lolex";
import { act, renderHook, cleanup } from "react-hooks-testing-library";
import { CountdownConfig } from "./countdown";

function getWrapper(): { wrapper: (props: any) => JSX.Element } {
  const timeSync = new TimeSync();
  return {
    wrapper: function TestWrapper(props: any) {
      return (
        <TimeContext.Provider
          value={{
            getCurrentTime: TimeSync.getCurrentTime,
            getTimeLeft: TimeSync.getTimeLeft,
            addTimer: timeSync.addTimer,
            createCountdown: timeSync.createCountdown
          }}
          {...props}
        />
      );
    }
  };
}

describe("#Countdown", () => {
  let clock: lolex.Clock;

  beforeEach(() => {
    clock = lolex.install({ now: 1 });
  });

  afterEach(() => {
    clock.uninstall();
    cleanup();
  });

  function actTicks(interval: number, count: number): void {
    for (let i = 0; i < count * 2; i++) {
      act(() => {
        clock.tick(interval / 2);
      });
    }
  }

  it("should be exported correctly", () => expect(useCountdown).toBeDefined());

  it("should throw if context is not found", () => {
    expect(() =>
      renderHook(() => useCountdown())
    ).toThrowErrorMatchingSnapshot();
  });

  it("should not update when until number is reached", () => {
    clock.tick(999);
    let renderCalledCount = 0;

    const { result, unmount } = renderHook(() => {
      renderCalledCount++;
      return useCountdown({ until: 1 });
    }, getWrapper());
    expect(result).toMatchSnapshot();
    actTicks(1000, 10);
    expect(result).toMatchSnapshot();
    act(() => {
      unmount();
    });
    expect(renderCalledCount).toBe(1);
  });

  it("should stop countdown if it has ended", () => {
    let renderCalledCount = 0;
    const timeLefts: number[] = [];

    const { result, unmount } = renderHook(() => {
      renderCalledCount++;
      const timeLeft = useCountdown({ until: 5002 });
      timeLefts.push(timeLeft);
      return timeLeft;
    }, getWrapper());
    expect(result).toMatchSnapshot();

    actTicks(1000, 6);
    expect(result).toMatchSnapshot();
    act(() => {
      unmount();
    });
    expect(renderCalledCount).toBe(7);
    expect(timeLefts).toEqual([6, 5, 4, 3, 2, 1, 0]);
  });

  it("should update countdown if props are updated", () => {
    let renderCalledCount = 0;
    const timeLefts: number[] = [];

    const countdownConfig: CountdownConfig = {};
    const { rerender } = renderHook(() => {
      renderCalledCount++;
      const timeLeft = useCountdown({ until: countdownConfig.until || 2001 });
      timeLefts.push(timeLeft);
      return timeLeft;
    }, getWrapper());

    actTicks(1000, 5);

    expect(renderCalledCount).toBe(3);
    countdownConfig.until = 15000;
    rerender();
    actTicks(1000, 20);
    expect(renderCalledCount).toBe(15);
    expect(timeLefts).toEqual([2, 1, 0, 10, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]);
  });

  it("should use the interval provided as a prop", () => {
    let renderCalledCount = 0;
    const timeLefts: number[] = [];

    renderHook(() => {
      renderCalledCount++;
      const timeLeft = useCountdown({
        until: 1000 * 60 * 60 * 2,
        interval: TimeSync.HOURS
      });
      timeLefts.push(timeLeft);
      return timeLeft;
    }, getWrapper());
    actTicks(1000 * 60 * 60, 5);
    expect(renderCalledCount).toBe(3);
    expect(timeLefts).toEqual([2, 1, 0]);
  });

  it("should not start a countdown if no until value is specified", () => {
    let renderCalledCount = 0;
    const timeLefts: number[] = [];

    renderHook(() => {
      renderCalledCount++;
      const timeLeft = useCountdown();
      timeLefts.push(timeLeft);
      return timeLeft;
    }, getWrapper());

    actTicks(1000, 10);

    expect(renderCalledCount).toBe(1);
    expect(timeLefts).toEqual([0]);
  });
});
