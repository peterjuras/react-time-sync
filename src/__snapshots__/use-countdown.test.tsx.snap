// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Countdown should not update when until number is reached 1`] = `
Object {
  "current": 0,
  "error": undefined,
}
`;

exports[`#Countdown should not update when until number is reached 2`] = `
Object {
  "current": 0,
  "error": undefined,
}
`;

exports[`#Countdown should stop countdown if it has ended 1`] = `
Object {
  "current": 6,
  "error": undefined,
}
`;

exports[`#Countdown should stop countdown if it has ended 2`] = `
Object {
  "current": 0,
  "error": undefined,
}
`;
